### PR TITLE
Use temporary db for benchmarking

### PR DESCRIPTION
**Pull request by michalkucharczyk**
If no db option was given benchmarks shall use temporary database. Otherwise the test can use locally stored database which maybe out-of-date causing test to fail.

### CODE REVIEW

**Comment by ggwpez** [comment]
> If no db option was given benchmarks shall use temporary database.  

This should happen automatically depending on whether a `dev` runtime or real runtime is used.  

> Otherwise the test can use locally stored database which maybe out-of-date causing test to fail.  

In which cases did you observe this?

File: utils/frame/benchmarking-cli/src/lib.rs [hunk 1]
```diff
@@ -87,6 +88,17 @@ impl CliConfiguration for BenchmarkCmd {
 		}
 	}
 
+	fn base_path(&self) -> Result<Option<BasePath>> {
+		let inner = unwrap_cmd! {
+			self, cmd, cmd.base_path()
+		};
+
+		match inner {
```
Comment (bkchr) [inline utils/frame/benchmarking-cli/src/lib.rs:96]: Some comment on why we are doing this would be nice.

**Comment by michalkucharczyk** [comment]
To reproduce the issue (assuming no local database stored):
Check out `polkadot` repo locally at 7222d32f2ad80a41b8df1dff735f6ed1a8d49f49.
Execute tests `cargo test --release --features=runtime-benchmarks` -- all good so far.

Then switch to 00fed3bc47dcb9cb591bdec1b2b2e703051241ae, and run tests again.
Tests should start to fail (it was my case).

The error message was:
```
...
2022-09-13 15:41:45 Running 1 warmups...    
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 Executing block 1 times    
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 Executing a balances::transfer_keep_alive extrinsic takes[ns]:
Total: 111840
Min: 111840, Max: 111840
Average: 111840, Median: 111840, Stddev: 0
Percentiles 99th, 95th, 75th: 111840, 111840, 111840    
2022-09-13 15:41:45 👶 Creating empty BABE epoch changes on what appears to be first startup.    
2022-09-13 15:41:45 Essential task `txpool-background` failed. Shutting down service.    
Error: 
   0: Extrinsic is not valid: TransactionValidityError::Invalid(InvalidTransaction::BadProof)
   1: Transaction has a bad signature

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 Running 1 warmups...    
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 Executing block 1 times    
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 ParentBlockRandomness did not provide entropy    
2022-09-13 15:41:45 Building block, this takes some time...    
test benchmark_extrinsic_works ... FAILED

failures:

---- benchmark_extrinsic_works stdout ----
thread 'benchmark_extrinsic_works' panicked at 'assertion failed: benchmark_extrinsic(&runtime, pallet, extrinsic).is_ok()', tests/benchmark_extrinsic.rs:31:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    benchmark_extrinsic_works

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 6.31s

error: test failed, to rerun pass '--test benchmark_extrinsic'

...
```

The single command line to reproduce the issue:
`cargo run  --features=runtime-benchmarks benchmark extrinsic --chain rococo-dev --pallet system --extrinsic remark --repeat=1 --warmup=1 --max-ext-per-block=1`
Adding `-d test-empty-db` parameter makes tests passing.

Also removal of `$HOME/.local/share/polkadot` makes the tests passing again (I am on Ubuntu).

**Comment by michalkucharczyk** [comment]
> This should happen automatically depending on whether a `dev` runtime or real runtime is used.

`dev` is not set to true in this case:
```
SharedParams { chain: Some("rococo-dev"), dev: false, base_path: None, log: [], detailed_log_output: false, disable_log_color: false, enable_log_reloading: false, tracing_targets: None, tracing_receiver: Log }
```

**Comment by ggwpez** [comment]
> he single command line to reproduce the issue: `cargo run --features=runtime-benchmarks benchmark extrinsic --chain rococo-dev`  

But then there is an issue with it not using a temporary directory for a `dev` runtime. IIUC `rococo-dev` should always use a temp directory, or not?   
This is then not a bug in benchmarking per-se.

**Comment by michalkucharczyk** [comment]
The parameters are modifed to `-dev` here:
https://github.com/paritytech/polkadot/blob/master/tests/benchmark_extrinsic.rs#L30

And later executed w/o `--dev`:
https://github.com/paritytech/polkadot/blob/master/tests/benchmark_extrinsic.rs#L45-L49

File: utils/frame/benchmarking-cli/src/lib.rs (hunk 1 above)
Comment (michalkucharczyk) [reply utils/frame/benchmarking-cli/src/lib.rs:98]: done

**Comment by michalkucharczyk** [comment]
> Okay that is a surprise for me. I was under the impression all `-dev` runtimes always imply `--dev`, since it cannot be passed explicit: `error: The argument '--chain <CHAIN_SPEC>' cannot be used with '--dev`. Is this expected behaviour @bkchr?

I think `--dev` option conflicts with `--chain` by design:
https://github.com/paritytech/substrate/blob/23bb5a6255bbcd7ce2999044710428bc4a7a924f/client/cli/src/params/shared_params.rs#L36-L39

From what I see the `-dev` suffix does not influence any of other parameters, the chain's name is not check for presence of `-dev`, e.g:
https://github.com/paritytech/polkadot/blob/2b0d83258089578b65712e143c794cf4922941de/cli/src/command.rs#L101

**Comment by bkchr** [comment]
> > `dev` is not set to true in this case:
> > ```
> > SharedParams { chain: Some("rococo-dev"), dev: false, base_path: None, log: [], detailed_log_output: false, disable_log_color: false, enable_log_reloading: false, tracing_targets: None, tracing_receiver: Log }
> > ```
> 
> Okay that is a surprise for me. I was under the impression all `-dev` runtimes always imply `--dev`, since it cannot be passed explicit: `error: The argument '--chain <CHAIN_SPEC>' cannot be used with '--dev`. Is this expected behaviour @bkchr?


Yes, sort of. `--dev` had always this same behavior. The assumption that there is only one `dev` chain. One of the fuckups of the cli. Maybe we should just get entirely rid of it.

**Review by ggwpez — APPROVED** [review]
> Yes, sort of. --dev had always this same behavior. The assumption that there is only one dev chain. One of the fuckups of the cli. Maybe we should just get entirely rid of it.

Yea either removing the `-dev` runtimes or the `--dev` flag would be nice.

**Comment by michalkucharczyk** [comment]
Not sure if doable, but how about extending `--dev` with chain name?
Default name could be `dev`.
e.g.
`--dev` -> `--chain=dev --force-authoring --rpc-cors=all ...` 
`--dev=polkadot-dev` -> `--chain=polkadot-dev --force-authoring --rpc-cors=all ...`

**Comment by ggwpez** [comment]
> Chain name given to `--dev` could be restricted to those ending with `-dev`. (Or alternatively `-dev` could be appended to the chain name provided).

Are `--dev` and `chain` not orthogonal to each other? I would rather delete the `-dev` runtimes.

**Comment by michalkucharczyk** [comment]
I may be wrong but `-dev` chains are about the chain parameters (authority set, wasm blob used).
`--dev` option forces some node's parameters.

File: utils/frame/benchmarking-cli/src/lib.rs [hunk 2]
```diff
@@ -87,6 +88,19 @@ impl CliConfiguration for BenchmarkCmd {
 		}
 	}
 
+	fn base_path(&self) -> Result<Option<BasePath>> {
+		let inner = unwrap_cmd! {
+			self, cmd, cmd.base_path()
+		};
+
+		//If the base path was not provided, benchmark command shall use temporary path. Otherwise
+		//we may end up using shared path, which may be inappropriate for benchmarking.
```
Comment (bkchr) [inline utils/frame/benchmarking-cli/src/lib.rs:97]:
```suggestion
		// If the base path was not provided, benchmark command shall use temporary path. Otherwise
		// we may end up using shared path, which may be inappropriate for benchmarking.
```

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge
